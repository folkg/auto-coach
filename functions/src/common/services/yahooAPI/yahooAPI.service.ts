import { AxiosError } from "axios";
import { logger } from "firebase-functions";
import { LineupChanges } from "../../../dispatchSetLineup/interfaces/LineupChanges";
import { PlayerTransaction } from "../../../dispatchSetLineup/interfaces/PlayerTransaction";
import { Token, YahooRefreshRequestBody } from "../../interfaces/credential";
import { updateFirestoreTimestamp } from "../firebase/firestore.service";
import {
  httpGetAxios,
  httpPostAxiosAuth,
  httpPostAxiosUnauth,
  httpPutAxios,
} from "./yahooHttp.service";
const js2xmlparser = require("js2xmlparser");

/**
 * Refresh the Yahoo access token for the given user
 * @param {string} refreshToken The refresh token
 * @return {Promise<Token>} The new credential
 */
export async function refreshYahooAccessToken(
  refreshToken: string
): Promise<Token> {
  const url = "https://api.login.yahoo.com/oauth2/get_token";
  const requestBody: YahooRefreshRequestBody = {
    client_id: process.env.YAHOO_CLIENT_ID as string,
    client_secret: process.env.YAHOO_CLIENT_SECRET as string,
    redirect_uri: process.env.YAHOO_REDIRECT_URI as string,
    refresh_token: refreshToken,
    grant_type: "refresh_token",
  };
  const body = Object.keys(requestBody)
    .map(
      (key) =>
        encodeURIComponent(key) +
        "=" +
        encodeURIComponent(requestBody[key as keyof YahooRefreshRequestBody])
    )
    .join("&");

  try {
    const requestTime = Date.now();
    const { data } = await httpPostAxiosUnauth(url, body);
    // Get the token info from the response and save it to the database
    const accessToken = data.access_token;
    const tokenExpirationTime = data.expires_in * 1000 + requestTime;
    const token: Token = {
      accessToken: accessToken,
      refreshToken: data.refresh_token,
      tokenExpirationTime: tokenExpirationTime,
    };

    return token;
  } catch (error: AxiosError | unknown) {
    handleAxiosError(error, null);
  }
}

/**
 * Will get the JSON response from Yahoo for all teams matching the TeamIDs
 * This will be useful if we want to get for just individual teams
 *
 * @async
 * @param {string} teams - comma separated string of teamIDs, ie.
 * "414.l.240994.t.12, 414.l.358976.t.4, 419.l.14950.t.2,
 * 419.l.19947.t.6,419.l.28340.t.1,419.l.59985.t.12"
 * @param {string} uid - The firebase uid
 * @param {string} [date=""] - The date to get the roster for. Defaults to today.
 * @return {Promise<any>} The Yahoo JSON object containing the rosters
 */
export async function getRostersByTeamID(
  teams: string[],
  uid: string,
  date = ""
): Promise<any> {
  const leagueKeysArray: string[] = [];
  teams.forEach((teamKey) => {
    const leagueKey = teamKey.split(".t")[0];
    if (!leagueKeysArray.includes(leagueKey)) {
      leagueKeysArray.push(leagueKey);
    }
  });
  const leagueKeys = leagueKeysArray.join(",");

  const url =
    "users;use_login=1/games;game_keys=nhl,nfl,nba,mlb" +
    "/leagues;league_keys=" +
    leagueKeys +
    ";out=settings/teams/roster;date=" +
    date +
    "/players;out=percent_started,percent_owned,ranks,opponent,starting_status" +
    ";ranks=last30days,last14days,projected_next7days,projected_season_remaining,last4weeks,projected_week,projected_next4weeks" +
    ";percent_started.cut_types=diamond" +
    ";percent_owned.cut_types=diamond" +
    "?format=json";

  try {
    const { data } = await httpGetAxios(url, uid);
    return data;
  } catch (err: AxiosError | unknown) {
    const errMessage = `Error in getRostersByTeamID. User: ${uid} Teams: ${teams}`;
    handleAxiosError(err, errMessage);
  }
}

/**
 * Will fetch the top 25 free agents for the given league, based on rank over
 * the last 14 days
 *
 * @export
 * @async
 * @param {string} leagueKey - The league key
 * @param {string} uid - The Yahoo user ID
 * @return {Promise<any>}
 */
export async function getFreeAgents(
  leagueKey: string,
  uid: string
): Promise<any> {
  // sort=AR_L30;sort_type=lastmonth
  const url =
    "users;use_login=1/games;game_keys=nhl,nfl,nba,mlb/leagues;league_keys=" +
    leagueKey +
    "/players;status=A;sort=AR_L14;sort_type=biweekly" +
    ";out=ownership,percent_started,percent_owned,ranks,opponent,starting_status" +
    ";ranks=last30days,last14days,projected_next7days,projected_season_remaining,last4weeks,projected_week,projected_next4weeks" +
    ";percent_started.cut_types=diamond" +
    ";percent_owned.cut_types=diamond" +
    "?format=json";

  try {
    const { data } = await httpGetAxios(url, uid);
    return data;
  } catch (err: AxiosError | unknown) {
    const errMessage = `Error in getFreeAgents. User: ${uid} League: ${leagueKey}`;
    handleAxiosError(err, errMessage);
  }
}

/**
 * Get all the standings for all the leagues for all the games from Yahoo
 * @async
 * @param {string} uid The firebase uid
 * @return {Promise<any>} The Yahoo JSON object containing league standings
 */
export async function getAllStandings(uid: string): Promise<any> {
  try {
    const { data } = await httpGetAxios(
      "users;use_login=1/games;game_keys=nfl,nhl,nba,mlb/" +
        "leagues/standings?format=json",
      uid
    );
    return data;
  } catch (err: AxiosError | unknown) {
    const errMessage = `Error in getAllStandings. User: ${uid}`;
    handleAxiosError(err, errMessage);
  }
}

/**
 * Get the projected/confirmed starting goalies for the league
 *
 * This will return starting goalies based on the weekly_deadline
 * ie. intraday leagues will return today's starters, while daily leagues
 * will return tomorrow's starters
 *
 * @export
 * @async
 * @param {string} uid The firebase uid
 * @param {string} leagueKey The league key
 * @return {Promise<any>} The Yahoo JSON object containing goalie data
 */
export async function getStartingGoalies(
  uid: string,
  leagueKey: string
): Promise<any> {
  try {
    // There could be up to 32 starting goalies, so we need to make 2 calls
    // to get all the goalies. The first call will get the first 25 goalies.
    const urlBase = `league/${leagueKey}/players;position=S_G;sort=AR;start=`;
    const [data1, data2] = await Promise.all([
      httpGetAxios(`${urlBase}0?format=json`, uid),
      httpGetAxios(`${urlBase}25?format=json`, uid),
    ]);
    return [data1.data, data2.data];
  } catch (err: AxiosError | unknown) {
    const errMessage = `Error in getStartingGoalies. User: ${uid} League: ${leagueKey}`;
    handleAxiosError(err, errMessage);
  }
}

/**
 * Post the roster changes to Yahoo
 *
 * @export
 * @async
 * @param {LineupChanges[]} lineupChanges
 * @param {string} uid The firebase uid of the user
 * @return {unknown}
 */
export async function putLineupChanges(
  lineupChanges: LineupChanges[],
  uid: string
): Promise<void> {
  for (const lineupChange of lineupChanges) {
    const { teamKey, coverageType, coveragePeriod, newPlayerPositions } =
      lineupChange;
    if (Object.keys(newPlayerPositions).length !== 0) {
      const players = [];
      for (const [playerKey, position] of Object.entries(newPlayerPositions)) {
        players.push({ player_key: playerKey, position });
      }

      const data = {
        roster: {
          coverage_type: coverageType,
          [coverageType]: coveragePeriod,
          players: { player: players },
        },
      };

      const XML_NAMESPACE = "fantasy_content";
      const xmlBody = js2xmlparser.parse(XML_NAMESPACE, data);

      try {
        await httpPutAxios(uid, `team/${teamKey}/roster?format=json`, xmlBody);
        logger.log(
          `Successfully put roster changes for team: ${teamKey} for user: ${uid}`
        );
        updateFirestoreTimestamp(uid, teamKey);
      } catch (err: AxiosError | unknown) {
        const errMessage = `Error in putLineupChanges. User: ${uid} Team: ${teamKey}`;
        handleAxiosError(err, errMessage);
      }
    } else {
      updateFirestoreTimestamp(uid, teamKey);
    }
  }
}

/**
 * Post the roster add/drop transactions to Yahoo
 *
 * @export
 * @async
 * @param {PlayerTransaction} transaction The roster transactions.
 * Shall contain the teamKey and the players to add/drop for a single transaction. This means that
 * the players array shall contain only 1 or 2 players.
 * @param {string} uid The Yahoo uid of the user
 * @return {Promise<void>}
 */
export async function postRosterAddDropTransaction(
  transaction: PlayerTransaction,
  uid: string
): Promise<boolean> {
  const { teamKey, players } = transaction;

  const validPlayerCount = [1, 2].includes(players.length);
  if (!validPlayerCount) {
    logger.warn(
      `Invalid number of players to move: ${players.length} for team: ${teamKey} for user: ${uid}. Must be 1 or 2.`
    );
    return false;
  }
  const XMLPlayers = players.map((player) => ({
    player_key: player.playerKey,
    transaction_data: {
      type: player.transactionType,
      [player.transactionType === "add"
        ? "destination_team_key"
        : "source_team_key"]: teamKey,
    },
  }));
  XMLPlayers.sort((a, b) =>
    a.transaction_data.type > b.transaction_data.type ? 1 : -1
  ); // sorts add before drop

  const transactionType =
    XMLPlayers.length === 1 ? XMLPlayers[0].transaction_data.type : "add/drop";

  const isWaiverClaim = false; // TODO: Implement waiver claim logic later
  const data = {
    transaction: {
      type: transactionType,
      ...(XMLPlayers.length === 1 && { player: XMLPlayers[0] }),
      ...(XMLPlayers.length === 2 && { players: { player: XMLPlayers } }),
      ...(isWaiverClaim ? { faab_bid: 0 } : {}),
    },
  };

  const XML_NAMESPACE = "fantasy_content";
  const xmlBody = js2xmlparser.parse(XML_NAMESPACE, data);

  const leagueKey = teamKey.split(".t")[0];
  try {
    await httpPostAxiosAuth(uid, `league/${leagueKey}/transactions`, xmlBody);
    logger.log(
      `Successfully posted ${transactionType} transaction for team: ${teamKey} for user: ${uid}.`
    );
    logger.log(`Transaction data: ${data}`);
    return true;
  } catch (err) {
    const errMessage = `Error in postRosterAddDropTransaction. User: ${uid} Team: ${teamKey}`;
    handleAxiosError(err, errMessage);
  }
}

/**
 * Handle the axios error
 *
 * @param {AxiosError} err - The axios error
 * @param {string} message - The message to throw
 */
function handleAxiosError(
  err: AxiosError | any,
  message: string | null
): never {
  const errMessage = message ? `${message}. ` : "";
  if (err.response) {
    throw new Error(
      `${errMessage}Error status: ${JSON.stringify(
        err.response.status
      )}: ${JSON.stringify(err.response.data)}`
    );
  } else if (err.request) {
    throw new Error(`${errMessage}${JSON.stringify(err.request)}`);
  } else {
    throw new Error(`${errMessage}${JSON.stringify(err.message)}`);
  }
}
