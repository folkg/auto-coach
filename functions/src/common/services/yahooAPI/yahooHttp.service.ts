import axios from "axios";
import { loadYahooAccessToken } from "../firebase/firestore.service";

const API_URL = "https://fantasysports.yahooapis.com/fantasy/v2/";

/**
 * Perform an HTTP put request to the yahoo API
 *
 * @export
 * @async
 * @param {string} url - the url to fetch
 * @param {string} uid The firebase uid of the user
 * @return {unknown} - the response from the API
 */
export async function httpGetAxios(url: string, uid: string) {
  const credential = await loadYahooAccessToken(uid);
  const accessToken = credential.accessToken;

  return axios.get(API_URL + url, {
    headers: {
      Authorization: "Bearer " + accessToken,
    },
  });
}

/**
 * Perform an HTTP post request to the Yahoo API
 *
 * @export
 * @async
 * @param {string} url - the FULL url to post to. Does not use API_URL.
 * @param {*} body - the body of the post request
 * @return {unknown} - the response from the API
 */
export function httpPostAxios(url: string, body: any) {
  return axios.post(url, body, {
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
    },
  });
}

/**
 * Perform an HTTP put request to the yahoo API
 *
 * @export
 * @async
 * @param {string} uid The firebase uid of the user
 * @param {string} url - the url to fetch
 * @param {string} body - the body of the put request
 * @return {unknown} - the response from the API
 */
export async function httpPutAxios(uid: string, url: string, body: string) {
  const credential = await loadYahooAccessToken(uid);
  const accessToken = credential.accessToken;

  return axios.put(API_URL + url, body, {
    headers: {
      "content-type": "application/xml; charset=UTF-8",
      Authorization: "Bearer " + accessToken,
    },
  });
}
