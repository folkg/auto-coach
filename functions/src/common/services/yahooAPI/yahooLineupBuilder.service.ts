import { IPlayer } from "../../interfaces/IPlayer.js";
import { ITeamOptimizer } from "../../interfaces/ITeam.js";
import {
  getChild,
  getPacificEndOfDay,
  getPacificStartOfDay,
  parseStringToInt,
} from "../utilities.service.js";
import { getRostersByTeamID } from "./yahooAPI.service.js";
import getPlayersFromRoster from "./yahooPlayerProcessing.service.js";
import {
  createTransactionArray,
  getGamesPlayedArray,
  getInningsPitchedArray,
  getPositionCounts,
} from "./yahooTeamProcesssing.services.js";

/**
 * Get the roster objects for the given teams
 *
 * @export
 * @async
 * @param {string[]} teamKeys The team keys
 * @param {string} uid The firebase uid of the user
 * @param {string} [date=""] The date to get the roster for. If not provided, the default "" is today's date
 * @return {Promise<ITeamOptimizer[]>} The roster objects
 */
export async function fetchRostersFromYahoo(
  teamKeys: string[],
  uid: string,
  date = ""
): Promise<ITeamOptimizer[]> {
  const result: ITeamOptimizer[] = [];

  if (teamKeys.length === 0) {
    return result;
  }

  const yahooJSON = await getRostersByTeamID(teamKeys, uid, date);
  const gamesJSON = getChild(yahooJSON.fantasy_content.users[0].user, "games");

  // Loop through each "game" (nfl, nhl, nba, mlb)
  for (const gameKey of Object.keys(gamesJSON).filter(
    (key) => key !== "count"
  )) {
    const gameJSON = gamesJSON[gameKey].game;
    const leaguesJSON = getChild(gameJSON, "leagues");

    // Loop through each league within the game
    for (const leagueKey of Object.keys(leaguesJSON).filter(
      (key) => key !== "count"
    )) {
      const league = leaguesJSON[leagueKey].league;
      const leagueSettings = getChild(league, "settings");
      const usersTeam = getChild(league, "teams")[0].team;
      const usersTeamRoster = getChild(usersTeam, "roster");
      const isEditable = usersTeamRoster.is_editable;
      if (!isEditable) {
        // skip this team if it is not editable
        continue;
      }
      const coverageType = usersTeamRoster.coverage_type;
      // save the position counts to a map
      const rosterPositions = getPositionCounts(leaguesJSON, leagueKey);
      const players: IPlayer[] = getPlayersFromRoster(
        usersTeamRoster[0].players
      );
      const gamesPlayedArray = getGamesPlayedArray(usersTeam);
      const inningsPitchedArray = getInningsPitchedArray(usersTeam);

      // TODO: Could add max/current adds condtionally as well
      const rosterObject: ITeamOptimizer = {
        team_key: getChild(usersTeam[0], "team_key"),
        team_name: getChild(usersTeam[0], "name"),
        league_name: getChild(league, "name"),
        players: players,
        coverage_type: coverageType,
        coverage_period: usersTeamRoster[coverageType],
        weekly_deadline: getChild(league, "weekly_deadline"),
        edit_key: getChild(league, "edit_key"),
        game_code: getChild(gameJSON, "code"),
        num_teams: getChild(league, "num_teams"),
        roster_positions: rosterPositions,
        scoring_type: getChild(usersTeam[0], "scoring_type"),
        current_weekly_adds: parseStringToInt(
          getChild(usersTeam[0], "roster_adds").value,
          0
        ),
        current_season_adds: parseStringToInt(
          getChild(usersTeam[0], "number_of_moves"),
          0
        ),
        max_weekly_adds: parseStringToInt(
          getChild(leagueSettings, "max_weekly_adds")
        ),
        max_season_adds: parseStringToInt(getChild(leagueSettings, "max_adds")),
        start_date: getPacificStartOfDay(getChild(league, "start_date")),
        end_date: getPacificEndOfDay(getChild(league, "end_date")),
        faab_balance: parseStringToInt(getChild(usersTeam[0], "faab_balance")),
        waiver_rule: getChild(leagueSettings, "waiver_rule"),
        transactions: createTransactionArray(
          getChild(usersTeam, "transactions")
        ),
        ...(gamesPlayedArray && { games_played: gamesPlayedArray }),
        ...(inningsPitchedArray && { innings_pitched: inningsPitchedArray }),
      };
      result.push(rosterObject);
    }
  }
  // logger.log("Fetched rosters from Yahoo API:");
  // console.log(JSON.stringify(result));
  return result;
}
