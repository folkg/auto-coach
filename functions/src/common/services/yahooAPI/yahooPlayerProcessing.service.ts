import type { IPlayer, PlayerRanks } from "../../interfaces/IPlayer.js";
import { getChild, parseStringToInt } from "../utilities.service.js";

/**
 * Deconstruct the players JSON object to get the required properties
 *
 * @param {*} playersJSON - The players JSON object
 * @param {Set<string>} [postponedTeams=new Set()] - A set of teams with postponed games today
 * @return {IPlayer[]} - An array of Player objects
 */
export default function getPlayersFromRoster(
  playersJSON: unknown,
  postponedTeams: Set<string> = new Set(),
): IPlayer[] {
  const result: IPlayer[] = [];

  // TODO: ArkType
  for (const key in playersJSON) {
    if (key !== "count") {
      const player = playersJSON[key].player;
      const opponent = getChild(player, "opponent");
      const selectedPosition = getChild(player, "selected_position");
      const teamId = getChild(player[0], "editorial_team_key");

      const playerObject: IPlayer = {
        player_key: getChild(player[0], "player_key"),
        player_name: getChild(player[0], "name").full,
        eligible_positions: getEligiblePositions(player),
        display_positions: getChild(player[0], "display_position").split(","),
        selected_position:
          selectedPosition && getChild(selectedPosition, "position"),
        is_editable: getChild(player, "is_editable") === 1,
        is_playing:
          !!opponent && opponent !== "Bye" && !postponedTeams.has(teamId),
        injury_status: getChild(player[0], "status_full") || "Healthy",
        percent_started: getPercentObject(player, "percent_started"),
        percent_owned: getPercentObject(player, "percent_owned"),
        percent_owned_delta: parseStringToInt(
          getChild(getChild(player, "percent_owned"), "delta"),
        ),
        is_starting: getChild(player, "starting_status")
          ? getChild(getChild(player, "starting_status"), "is_starting")
          : "N/A",
        ranks: getPlayerRanks(player),
        is_undroppable: getChild(player[0], "is_undroppable") === "1",
      };

      const ownership = getChild(player, "ownership");
      if (ownership) {
        playerObject.ownership = ownership;
      }

      result.push(playerObject);
    }
  }

  return result;
}

function getEligiblePositions(player: unknown) {
  const eligiblePositions: string[] = [];
  // TODO: ArkType
  for (const position of getChild(player[0], "eligible_positions")) {
    eligiblePositions.push(position.position);
  }
  return eligiblePositions;
}

/**
 * Will get the diamond cut value for the percent started or percent owned
 *
 * @param {*} player - The player JSON object
 * @param {string} percentType - The type of percent to get
 * (percent_started or percent_owned)
 * @param {string} cut - The cut type to get (diamond, platinum, gold, silver, bronze)
 * @return {number} - The diamond cut value
 */
function getPercentObject(
  player: unknown,
  percentType: string,
  cut = "diamond",
): number {
  // TODO: ArkType
  const percentObject = getChild(player, percentType);
  let result = getChild(percentObject, "value");

  // if we can get the cut type, then we will return that instead of the general value
  const percentCuts = getChild(percentObject, `${percentType}_cut_types`);
  for (const cutType of percentCuts) {
    const cutTypeObject = cutType[`${percentType}_cut_type`];
    if (getChild(cutTypeObject, "cut_type") === cut) {
      result = getChild(cutTypeObject, "value");
    }
  }

  return Number(result ?? 0);
}

/**
 * Will get the player ranks for the player
 *
 * @param {*} player - The player JSON object
 * @return {*} - The player ranks
 */
function getPlayerRanks(player: unknown): PlayerRanks {
  const rankTypeMap: { [key: string]: keyof PlayerRanks } = {
    L30: "last30Days",
    L14: "last14Days",
    PS7: "next7Days",
    PSR: "restOfSeason",
    L4W: "last4Weeks",
    PW: "projectedWeek",
    PN4W: "next4Weeks",
  };
  const result: PlayerRanks = {
    last30Days: -1,
    last14Days: -1,
    next7Days: -1,
    restOfSeason: -1,
    last4Weeks: -1,
    projectedWeek: -1,
    next4Weeks: -1,
  };

  // TODO: ArkType
  const ranks = getChild(player, "player_ranks");
  for (const rank of ranks) {
    const rankType = rank.player_rank.rank_type;
    if (rankType in rankTypeMap) {
      const key = rankTypeMap[rankType];
      result[key] = parseStringToInt(rank.player_rank.rank_value);
    }
  }

  return result;
}
