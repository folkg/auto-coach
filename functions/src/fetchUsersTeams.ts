import { HttpsError, onCall } from "firebase-functions/v2/https";
import { TeamClient, TeamFirestore } from "./interfaces/team";
import { fetchTeamsYahoo } from "./services/fetchUsersTeams.service";
import {
  fetchTeamsFirestore,
  syncTeamsInFirebase,
} from "./services/firestore.service";

export const fetchuserteams = onCall(async (request): Promise<TeamClient[]> => {
  const uid = request.auth?.uid;
  if (!uid) {
    throw new HttpsError(
      "unauthenticated",
      "You must be logged in to get an access token"
    );
  }

  const [yahooTeams, firestoreTeams] = await Promise.all([
    fetchTeamsYahoo(uid),
    fetchTeamsFirestore(uid),
  ]).catch((err: Error | any) => {
    throw new HttpsError("internal", err);
  });

  if (yahooTeams.length === 0) {
    throw new HttpsError(
      "internal",
      "No teams were returned from Yahoo. Please try again later."
    );
  }

  // Update the teams in firestore if required
  syncTeamsInFirebase(yahooTeams, uid, firestoreTeams);

  const teams: TeamClient[] = [];
  firestoreTeams.forEach((firestoreTeam: TeamFirestore) => {
    const yahooTeam: TeamClient | undefined = yahooTeams.find(
      (t) => t.team_key === firestoreTeam.team_key
    );
    if (yahooTeam) {
      // remove the team from the yahooTeams array and merge it with the team from firebase
      yahooTeams.splice(yahooTeams.indexOf(yahooTeam), 1);
      teams.push({ ...yahooTeam, ...firestoreTeam });
    }
  });
  // add the remaining teams from yahoo to the teams array for display on the frontend
  teams.push(...yahooTeams);

  return teams;
});
