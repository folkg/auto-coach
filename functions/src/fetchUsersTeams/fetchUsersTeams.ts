import { HttpsError, onCall } from "firebase-functions/v2/https";
import { TeamClient, TeamFirestore } from "../common/interfaces/team";
import { fetchTeamsYahoo } from "./services/fetchUsersTeams.service";
import {
  fetchTeamsFirestore,
  syncTeamsInFirebase,
} from "../common/services/firebase/firestore.service";

export const fetchuserteams = onCall(async (request): Promise<TeamClient[]> => {
  const uid = request.auth?.uid;
  if (!uid) {
    throw new HttpsError(
      "unauthenticated",
      "You must be logged in to get an access token"
    );
  }

  const [yahooTeams, firestoreTeams] = await Promise.all([
    fetchTeamsYahoo(uid),
    fetchTeamsFirestore(uid),
  ]).catch((err: Error | any) => {
    throw new HttpsError("data-loss", err.message);
  });

  if (yahooTeams.length === 0) {
    throw new HttpsError(
      "internal",
      "No teams were returned from Yahoo. Please try again later."
    );
  }

  // find all teams that are in firestore but not in yahoo
  const extraTeams = firestoreTeams.filter(
    (t) => !yahooTeams.some((y) => y.team_key === t.team_key)
  );

  const teams: TeamClient[] = [];
  firestoreTeams.forEach((firestoreTeam: TeamFirestore) => {
    const yahooTeam: TeamClient | undefined = yahooTeams.find(
      (t) => t.team_key === firestoreTeam.team_key
    );
    if (yahooTeam) {
      // remove the team from the yahooTeams array and merge it with the team from firebase
      yahooTeams.splice(yahooTeams.indexOf(yahooTeam), 1);
      teams.push({ ...yahooTeam, ...firestoreTeam });
    }
  });
  // add the remaining teams from yahoo to the teams array for display on the frontend
  teams.push(...yahooTeams);

  // Update the teams in firestore if required
  try {
    await syncTeamsInFirebase(yahooTeams, extraTeams, uid);
  } catch (err: Error | any) {
    console.log("Error syncing teams in firebase: " + err.message);
  }

  return teams;
});
