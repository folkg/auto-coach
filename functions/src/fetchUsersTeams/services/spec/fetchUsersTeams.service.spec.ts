import { fetchTeamsYahoo } from "../fetchUsersTeams.service";

jest.mock("firebase-admin", () => ({
  initializeApp: jest.fn(),
  firestore: jest.fn(),
}));
const mockGetAllStandings = jest.spyOn(
  require("../../../common/services/yahooAPI/yahooAPI.service"),
  "getAllStandings"
);

describe("fetchUsersTeams", () => {
  it("should return a list of teams", async () => {
    const input = require("./input.json");
    const expectedOutput = require("./output.json");
    mockGetAllStandings.mockReturnValue(input);

    const teams = await fetchTeamsYahoo("test");

    expect(teams).toEqual(expectedOutput);
  });
});
