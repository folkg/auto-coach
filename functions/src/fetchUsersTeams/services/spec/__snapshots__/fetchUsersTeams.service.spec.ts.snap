// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchUsersTeams > should return a list of teams v2 1`] = `
[
  {
    "current_season_adds": 20,
    "current_weekly_adds": 2,
    "edit_key": "2023-04-14",
    "end_date": 1681541999999,
    "faab_balance": -1,
    "game_code": "nhl",
    "game_is_over": 1,
    "game_name": "Hockey",
    "game_season": "2022",
    "league_name": "League Name",
    "max_games_played": -1,
    "max_innings_pitched": -1,
    "max_season_adds": 60,
    "max_weekly_adds": -1,
    "num_teams": 12,
    "outcome_totals": undefined,
    "points_against": undefined,
    "points_back": 51.5,
    "points_for": "65.5",
    "rank": "9",
    "roster_positions": {
      "BN": 4,
      "C": 2,
      "D": 4,
      "G": 2,
      "IR": 2,
      "IR+": 2,
      "LW": 2,
      "RW": 2,
      "Util": 2,
      "W": 1,
    },
    "scoring_type": "roto",
    "start_date": 1665126000000,
    "team_key": "419.l.14950.t.2",
    "team_logo": "https://yahoofantasysports-res.cloudinary.com/image/upload/t_s192sq/fantasy-logos/697b480fb677ab0cf351dd5413ce131c118c101d6c31574714bf6e2e8e7431fb.jpg",
    "team_name": "Bread Roller",
    "team_url": "https://hockey.fantasysports.yahoo.com/hockey/14950/2",
    "waiver_rule": "all",
    "weekly_deadline": "",
  },
  {
    "current_season_adds": 31,
    "current_weekly_adds": 0,
    "edit_key": "2023-04-02",
    "end_date": 1680505199999,
    "faab_balance": -1,
    "game_code": "nhl",
    "game_is_over": 1,
    "game_name": "Hockey",
    "game_season": "2022",
    "league_name": "Irish Lacrosse",
    "max_games_played": -1,
    "max_innings_pitched": -1,
    "max_season_adds": 50,
    "max_weekly_adds": 7,
    "num_teams": 12,
    "outcome_totals": {
      "losses": "70",
      "percentage": ".563",
      "ties": "25",
      "wins": "94",
    },
    "points_against": undefined,
    "points_back": undefined,
    "points_for": undefined,
    "rank": 1,
    "roster_positions": {
      "BN": 7,
      "C": 2,
      "D": 4,
      "G": 2,
      "IR": 2,
      "IR+": 1,
      "LW": 2,
      "NA": 2,
      "RW": 2,
    },
    "scoring_type": "head",
    "start_date": 1665126000000,
    "team_key": "419.l.19947.t.6",
    "team_logo": "https://yahoofantasysports-res.cloudinary.com/image/upload/t_s192sq/fantasy-logos/560d006423ab81292ec70a9db602d5ef73bce4c1e65d2553ef8772a34158b9dd.jpg",
    "team_name": "Gudas Hell",
    "team_url": "https://hockey.fantasysports.yahoo.com/hockey/19947/6",
    "waiver_rule": "all",
    "weekly_deadline": "intraday",
  },
  {
    "current_season_adds": 33,
    "current_weekly_adds": 0,
    "edit_key": "2023-04-14",
    "end_date": 1681541999999,
    "faab_balance": -1,
    "game_code": "nhl",
    "game_is_over": 1,
    "game_name": "Hockey",
    "game_season": "2022",
    "league_name": "Elite League 22",
    "max_games_played": -1,
    "max_innings_pitched": -1,
    "max_season_adds": 70,
    "max_weekly_adds": -1,
    "num_teams": 12,
    "outcome_totals": undefined,
    "points_against": undefined,
    "points_back": 24,
    "points_for": "90",
    "rank": "3",
    "roster_positions": {
      "BN": 5,
      "C": 2,
      "D": 4,
      "G": 2,
      "IR": 1,
      "IR+": 2,
      "LW": 2,
      "RW": 2,
      "Util": 3,
    },
    "scoring_type": "roto",
    "start_date": 1665126000000,
    "team_key": "419.l.28340.t.1",
    "team_logo": "https://s.yimg.com/cv/apiv2/default/nhl/nhl_7.png",
    "team_name": "Secret AdMeierer",
    "team_url": "https://hockey.fantasysports.yahoo.com/hockey/28340/1",
    "waiver_rule": "all",
    "weekly_deadline": "",
  },
  {
    "current_season_adds": 12,
    "current_weekly_adds": 0,
    "edit_key": "2023-04-02",
    "end_date": 1680505199999,
    "faab_balance": -1,
    "game_code": "nhl",
    "game_is_over": 1,
    "game_name": "Hockey",
    "game_season": "2022",
    "league_name": "Brewhouse Appreciation Society",
    "max_games_played": -1,
    "max_innings_pitched": -1,
    "max_season_adds": -1,
    "max_weekly_adds": 3,
    "num_teams": 12,
    "outcome_totals": {
      "losses": "95",
      "percentage": ".483",
      "ties": "27",
      "wins": "88",
    },
    "points_against": undefined,
    "points_back": undefined,
    "points_for": undefined,
    "rank": 10,
    "roster_positions": {
      "BN": 4,
      "D": 4,
      "F": 6,
      "G": 2,
      "IR+": 2,
      "Util": 3,
    },
    "scoring_type": "head",
    "start_date": 1665126000000,
    "team_key": "419.l.59985.t.12",
    "team_logo": "https://yahoofantasysports-res.cloudinary.com/image/upload/t_s192sq/fantasy-logos/b0bf7ede474e024dbf514fdc9272582b4541220886e436e5bfff3a1cc3f8a4b4.jpg",
    "team_name": "Time and Tye McGinn",
    "team_url": "https://hockey.fantasysports.yahoo.com/hockey/59985/12",
    "waiver_rule": "all",
    "weekly_deadline": "",
  },
]
`;
